### PR TITLE
Remove HTML format from io.ascii guess list

### PR DESCRIPTION
**Pull request by taldcroft**
From #3691 it appears that the HTML-format reader (with the libxml backend for beautifulsoup) can segfault for normal CSV-format input. 

I would suggest mitigating this by taking HTML out of the standard guess list.  It is a bit of an outlier and not something that would commonly appear in a data file of unknown format.

@hamogu @ChrisBeaumont ?

### CODE REVIEW

**Comment by taldcroft** [comment]
PR attached.

**Comment by hamogu** [comment]
-1 on this. It might break people's code who right now rely on guessing of HTML tables. In the mid term fast_identify is a lot more promising and we should aim to do that before 1.1 (at least for HTML tables), so that this code would never be released.

In the short term, I suggest that @chrisbeaumont just monkey patches the guess list in his use case to look as suggested in this PR. Would that work?

**Comment by astrofrog** [comment]
I thought I made this comment somewhere but can't find it - this will not affect people using Table.read since I think that is done based on the extension for HTML?

In any case, as a workaround, we could also make ascii.read behave this way - i.e. disable the guessing as suggested by this PR, but add a special case that if the extension is htm or html, it will actually try the HTML parser and try it first? This will be backward-compatible for 99% of cases.

**Comment by embray** [comment]
What happened to the idea of pre-scanning for presence of a `<` before passing anything to the HTML parser?

**Comment by taldcroft** [comment]
@hamogu - to my mind this is a critical bug since it results in segfault and has been reported by at least 5 users.  Would you be at least +0 if we did the suggestion from @astrofrog to reduce (to a very low level) the breakage?  With his suggestion in place the only thing that would break is if you are passing some HTML table input besides a file name and asking for that to be guessed.

@embray - the problem with pre-scanning is that it still leaves open the loophole of a CSV file that happens to contain valid HTML.  If you lower the bar as low as just a `<` then a false positive is quite likely and the segfault can still appear.

**Comment by hamogu** [comment]
OK, convinced, but we might still want to reverse this once we have improved the guessing with a pretty scanning step.

**Comment by eteq** [comment]
:+1: to @astrofrog's suggestion on this.  

I should also note that I've had 2 other people specifically come to me with this problem within the last couple weeks (both are not active on github but it caused them significant problems which they might not have solved w/o me to ask...), so I agree with @taldcroft that this is critical. (Perhaps we should also get this in for 1.0.3, @embray ?)

**Comment by taldcroft** [comment]
I'll try to implement the suggestion for `io.ascii.read()`.

**Comment by taldcroft** [comment]
OK, maybe we have an acceptable solution to this issue.  The default now is to not guess HTML, **but** if the file is "probably" HTML then the HTML reader is put at the head of the guess list.  Please check the code for what is meant by "probably" and let me know if that seems fine.

The idea is to set the bar relatively high for "probably" to avoid false positives with randomly-occurring HTML-like characteristics, but still get most real HTML table inputs for guessing.  Nothing is changed in the HTML reader itself.

I think this will result in a small enough change to the behavior in 99% of real use cases that we could put this into a bug fix release, but comments welcome.

**Comment by hamogu** [comment]
Looks good to me (if the tests pass).

File: astropy/io/ascii/ui.py [hunk 1]
```diff
@@ -42,6 +42,51 @@
...
+            table = os.linesep.join(table[:i+1])
+        except:
+            pass
+
+    if isinstance(table, six.string_types):
+        # Look for signs of an HTML table in the first maxchars characters
+        table = table[:maxchars]
+
+        # URL ending in .htm or .html
+        if re.match(r'http[s]?:// .+ \.htm[l]?$', table, re.IGNORECASE):
```
Comment (mdboom) [inline astropy/io/ascii/ui.py]: I think this one needs `re.VERBOSE` in the flags also, since it includes spaces for clarity.

Comment (mdboom) [inline astropy/io/ascii/ui.py]: Also, I think should have `^` at beginning so it won't match on bogus protocols like `Xhttp://`.  Should probably also accept `ftp` and `file`.  Or maybe just `[a-z]{1-9}` for the protocol.

**Comment by embray** [comment]
Okay, I'll merge for v1.0.3, but note the changelog entry is misplaced, in that case.  I will move it after merging.

**Pull request merged**